### DISCUSSION TITLE
Provide a way to construct filters programatically using a builder pattern

### DISCUSSION THREAD

**Discussion by laeubi** [osgi/osgi#541] [Ideas]
`org.osgi.framework.FrameworkUtil`  has method `createFilter(String)` that throws `InvalidSyntaxException` because there is no way to know if the filter is not malformed.

It would be great to have something like

https://nightlies.apache.org/directory/api/1.0.2/apidocs/org/apache/directory/ldap/client/api/search/FilterBuilder.html

as an alternative to build a string using a builder-pattern instead of passing a raw string and need to build it manually.

This even would not require to parse the string again, because the builder can build the required objects internally.

**Comment by stbischof** [comment]
Could you describe a usecase where you would use this?

Not sure is this should live in the framework or util spec. Or just wrapping the apache directory ldap builder with an filter inteface

**Comment by laeubi** [comment]
Just everywhere where one can pass a filter (or string) in the framework to filter services or match things against a set of properties.

**Reply by HannesWell** [reply]
One example is `org.eclipse.osgi.service.datalocation.Location` that is for example used extensively in Eclipse `InternalPlatform` class:
https://github.com/eclipse-platform/eclipse.platform/blob/master/runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/InternalPlatform.java#L651-L677

But one can easily find other examples when for example searching the callers of the `ServiceTracker` constructor `ServiceTracker(BundleContext, Filter , ServiceTrackerCustomizer<S, T>)` in the environment of choice.

**Reply by tjwatson** [reply]
I don't think this is a good example to justify adding new spec API here.  The filters here are hard-coded, well tested already and never change.  The overhead of using a FilterBuilder in this case is like using a StringBuilder to build up a static final string vs just quoting the complete string in a single final constant that can get compiled into the class vs built up at runtime.

FilterBuilder is useful for cases that need to dynamically create filters based on some external input.  I believe that usecase impacts a relatively small group of developers.

**Reply by HannesWell** [reply]
Then maybe org.eclipse.core.internal.runtime.Activator is a better one?
There the filter is created with the protocol on each call:
https://github.com/eclipse-equinox/equinox/blob/master/bundles/org.eclipse.equinox.common/src/org/eclipse/core/internal/runtime/Activator.java#L271-L279

But for me this is not mainly about performance but about improving the readability of the code and simplicity to construct new ones. Although you are right that one usually does not create new ones very often (at least I don't).

**Reply by tjwatson** [reply]
Perhaps, but if I was to rewrite that code I would opt to use a single `ServiceTracker<URLConverter, ServiceReference<URLConverter>>` 
 with a simple filter `"(objectClass=" + URLConverter.class.getName() + ")"` and that tracker maintains a `Map<String, List<ServiceReference<URLConverter>>` where the key is the protocol and the list is a service rank ordered list of converters for that protocol.  Then when a specific protocol was requested it would call `ctx.getService(ServiceReference)` with the highest ranked reference for the protocol from the map.

That seems better to me than maintaining a map of trackers based on user input for protocols that may never exist.

**Reply by laeubi** [reply]
I really doubt that performance is an issue here, but for sure if that's an impact users can still use "faster" Strings.

Also of course one can always argue that it is constant in some way and testes, but the same way one can argue against static final constants as they are constant why extract them and thing about a good place/name and so on if they are constant and well testet?

It is also about readability and maintainability and generally I'm fine spending a few microsecconds for this. A good example is this one:

https://github.com/eclipse-equinox/p2/blob/55bc99b583490aec6a703ed34baf39d489f84e96/bundles/org.eclipse.equinox.p2.touchpoint.eclipse/src/org/eclipse/equinox/internal/p2/touchpoint/eclipse/LazyManipulator.java#L28-L33

this is really hard to read, hard to understand and horrible to maintain and shows the usual anti-pattern of catch/ignore exception because we know it could NEVER happen

> } catch (InvalidSyntaxException e) {
> 	// Can't happen we are writing the filter ourselves
> 	return null;
> } 

the bad thing is just if someone accidentally mess up the filter a NPE is thrown somewhere (or maybe nothing happens at all if code thinks null mean missing service) ...

**Reply by tjwatson** [reply]
> the bad thing is just if someone accidentally mess up the filter a NPE is thrown somewhere (or maybe nothing happens at all if code thinks null mean missing service) ...

To solve that I suggest we consider changing `InvalidSytaxException` to extend `RuntimeException`.

**Comment by HannesWell** [comment]
There is also a similar method in `BundleContext.createFilter(String)` that returns a Filter object.
As suggested in todays spec call for both `FrameworkUtil` and `BundleContext` there could be another method `createFilterBuilder()` that returns a `FilterBuilder` impl provided and trusted by the used Framework.

**Comment by HannesWell** [comment]
Because it was discussed in the Spec-Call today as well about throwing an `InvalidSyntaxException`.
When using a builder such exception actually should never be thrown, at least for those cases where one cannot make invalid inputs (or then instead a IllegalArgumentException or a NPE could also fit).

It was also argued that even syntactically valid filters can be semantically incorrect and therefore still cause exceptions to be thrown. But from my understanding those exceptions would then not be `InvalidSyntaxException` thrown by `createFilter()` but some other exception thrown by the receiver of the created filter object.
Therefore it looks valid to me that a `FilterBuilder` would not throw a `InvalidSyntaxException` although it can create semantically invalid filter, but they should always be syntactically correct.

**Reply by laeubi** [reply]
What is a "semantically" invalid filter? A Filter either match or do not match and I'm not aware of any assertion regarding the "content" of the filter.

**Comment by tjwatson** [comment]
Strings are immutable, builders are not. We did not go into details on overloading core APIs to take a FilterBuilder where Strings are used today (e.g. BundleContext.addServiceListener(ServiceListener, String)). Careful consideration will be needed when specifying how a framework implementation uses the mutable object if we are to pass in FilterBuilder objects to core APIs and what are the expected behaviors if a builder changes after it has been given to the framework.

**Reply by laeubi** [reply]
> Strings are immutable, builders are not. 

While the Builder is mutable, it is expected that a user would need to call e.g. 
```
Filter myFilter = createFilter().eq(Constants.OBJECT_CLASS, "test").build();
```
and then use the Filter object (that is immutable) to pass it along.

> We did not go into details on overloading core APIs to take a FilterBuilder where Strings are used today (e.g. BundleContext.addServiceListener(ServiceListener, String)).

I always wondered why there is not a method `BundleContext.addServiceListener(ServiceListener, Filter))` as it seems obvious like with `ServiceTracker`...

**Reply by tjwatson** [reply]
> I always wondered why there is not a method `BundleContext.addServiceListener(ServiceListener, Filter))` as it seems obvious like with `ServiceTracker`...

There are two reasons that we discussed in the spec call relating to accepting foreign implementations of the Filter implementation :

1) Framework can depend on always using its own Filter implementation and make optimizations for service event delivery and lookup
2) Framework can trust its own implementation to do the right thing

`FrameworkUtil.createFilter` was a mistake in my opinion.  The only core API where a `Filter` is passed is the constructor of ServiceTracker.  But the actual implementation of the `Filter` is not used for any match at all.  Instead it calls t`oString` and then that is parsed again by the framework to ensure it has its own impl of `Filter`.  For historical reference, `FrameworkUtil.createFilter` was added so the implementation of the various framework permissions could create `Filter` objects in their implementations.  But we could have instead made that `Filter` impl used for permission implementations to be a package private class to parse the filter string input for the permission implementation.  Then when we finally gut all permissions because no supported Java will have any ability to have a security manager then we could simply remove this non-API `Filter` impl from the spec classes.  But instead we are left with a public facing `createFilter` util method for all time.  I am not convinced adding yet another way to create a `Filter` object is a good idea.

**Reply by laeubi** [reply]
Even though its seldom, when needed Filters are really usefull, and it seems odd especially for ServiceTracker and alike not to simply call "match" on the provided filters...

**Reply by tjwatson** [reply]
IIRC, ServiceTracker with a Filter used to call match on the provided filter and then register a listener with no filter with the framework.  This side-stepped any optimizations the framework could do to avoid full filter evaluation against every service registered with the framework during tracker open or during any service event.  It also had issues because service listener hooks wanted more information on what services a bundle requires by looking at https://docs.osgi.org/javadoc/osgi.core/8.0.0/org/osgi/framework/hooks/service/ListenerHook.ListenerInfo.html#getFilter--

Before the ServiceTracker impl used no filter registered with the framework, resulting in ListenerHook.getFilter returning null.  Some scenarios needed to have more information on what service properties a bundle required.  @bosschaert I think that was important for stuff like remote services, correct?

**Reply by bosschaert** [reply]
Correct, remote service want to know what services are being looked for so that they can be discovered and registered on-demand.

**Comment by maho7791** [comment]
I like the builder idea in general. But I also understand Toms and BJ's arguments, not having that in the framework.
Wouldn't it an idea to establish a library like we have it for osgi-test or osgi-slf4j? There could be a small library project that provides such helpers without specification. Then the framework can stay as it is.

**Reply by laeubi** [reply]
The main problem with such "small helpers" is that it tends to be not used because it is "non-standard" and people have to fiddle with licensing deployment and release issues, while something provided by default is likely to be used more.

Beside that there are the following issues:

1. Either we must fall back to strings again that are then pared again to a filter or the framework must be happy to accept *any* `Filter` implementations an not only framework provided ones
2. Actually efforts are duplicated, this library must implement the filter syntax, and the framework must implement the syntax e.g look at Equinox [FilterImpl](https://github.com/eclipse-equinox/equinox/blob/master/bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/FilterImpl.java) it actually has object representations of an/or/eq ... so this could be reused for a Builder, actually the `Parser` **is** already a builder that works an a stream of characters as input instead of method calls.

What I think could work would be that there is one `FilterFactory` maintained and provided by the osgi-specification itself and equinox/felix/... could simply use this, actually all Filter handling is not OSGi Framework specific and not bound to the framework istself ... Equinox effectivly always delegates to `FilterImpl.<somestaticmethod>` anyways...

**Reply by maho7791** [reply]
It is just an idea. That things are not used because they are "non-standard" is just a hypothesis. 
People also tend to use tons of jars from Maven Central and very often don't care of they just use one class of a JAR.
Then I follow BJ's suggestion to provide a design document.

**Reply by HannesWell** [reply]
> It is just an idea. That things are not used because they are "non-standard" is just a hypothesis. People also tend to use tons of jars from Maven Central and very often don't care of they just use one class of a JAR.

From my limited experience it depends and the behavior is not consistent. Some do really care about 'official' dependencies and some don't.

**Comment by HannesWell** [comment]
Btw. is there a reason why `BundleContext.getServiceReference()` and its derivatives only supports String filter parameters and has no overload tos that accept Filter objects?
In case Filter objects are leveraged better, it could be considered to add that.

**Reply by bjhargrave** [reply]
I think the concern is foreign Filter impls which cannot be trusted and are not optimized for the framework impl. The string format of a filter is the serialized form of a Filter and is the best and most basic exchange format. The receiver of the string filter is then free to deserialize it as they wish.

**Reply by HannesWell** [reply]
That's right. But in that case wouldn't it be possible, with the approach suggested already that the Framework creates the Filter-Builder/Factory, that receivers of such strings check the actual class of the filter object and throw a `IllegalArgumentException` (or alike) if it is not the Frameworks implementation? There could be a Framework internal method like `requireEquinoxFilter()` that does that, just like there is for example `Objects.requireNonNull()`.

**Reply by bjhargrave** [reply]
But why this unnecessary complication? The whole idea of a builder for filters is a library issue not a specification issue. I don't see this as a productive use of the project's time. The only useful thing from this discussion would be to rebase InvalidSyntaxException on RuntimeException to simplify code authoring around calling methods which are documented to throw InvalidSyntaxException. Such a change is simple, backwards-compatible, and doesn't bloat the API with overloads and a builder type. It is easy to explain in the specification (it basically requires zero changes to the specification text).

I am a strong -1 on adding a builder type and overloading methods to take Filter objects.

**Reply by HannesWell** [reply]
Instead of specifying it in the OSGi spec we could also incubate that for example in Equinox (and/or others) and see if it works well and if it is used. In Eclipse-Platform we have a three years deprecation/removal policy (and if that applies to Equinox as well), we could remove it again if it goes really bad or if it goes very well, it can evolve to be part of the OSGi spec.
This would probably also be less work in the first step because we can start with the technical part first and have some experience when it comes to specifying it.

**Comment by laeubi** [comment]
Just curious, can someone explain why a filter must be "trusted"? Can I do anything bad with a custom filter impl?

Independent of this I think one could enhance @HannesWell idea in a simpler way by havein a method (e.g. for the equinox case) the following way:

```
FilterImpl getInternal(Filter filter) {
  if (filter instanceof FilterImpl) {
    return filter;
  }
  return FilterImpl.parse(filter.toString());
}
```
This will always return a "good" framework managed filter but prevents parsing the filter again if it is already of the desired type.

Passing strings with semantics is really bad as there is only runtime-checking and no compile-time (or even IDE-typing-time) checks so if a string is not a string (=some text without semantics that needs be parsed) I always would prefer an Object with an (interface) type. A good example is from the context of m2e/maven where the GAV is a string and there are a lots of constructors like

`DefaultArtifact(String groupId, String artifactId,  String version, String type, String classifier)`

this is a complete mess as one can easily mess up thing by e.g. accidentally flipping `type` and `classifier` or passing a wrong formatted version and so on, but the IDE/compiler can not help anything here.

**Reply by tjwatson** [reply]
> This will always return a "good" framework managed filter but prevents parsing the filter again if it is already of the desired type.

Why allow foreign filters to be passed in if they are not going to be called?  I realize we are in that situation with `ServiceTracker`, but I would not think we should repeat that mistake.

**Reply by laeubi** [reply]
I think it must not be allowed, but if one absolutely want to be "safe" it would be an option, but `IllegalArgumentException` would be fair enough as well.

**Reply by tjwatson** [reply]
> Passing strings with semantics is really bad as there is only runtime-checking and no compile-time (or even IDE-typing-time) checks so if a string is not a string (=some text without semantics that needs be parsed) I always would prefer an Object with an (interface) type.

If `InvalidSyntaxException` is changed to a `RuntimeException` and a builder is available (perhaps not from core) that only produces a syntactically correct filter `String` then it seems to solve the concern without overloading methods that take `String` filters to also take `Filter` objects.

**Comment by laeubi** [comment]
By the way I don't think we need a dedicated "Builder" type but can use the functional style of recent java API, e.g. assume the Filterinterface is enhanced with
<details>
  <summary> these methods (framework impl might override for better performance)</summary>

```
/**
 * Creates a Filter that is the negation of the current instance
 * 
 * @return this filter negated
 */
default Filter not() {
	try {
		return FrameworkUtil.createFilter("(!" + toString() + ")");
	} catch (InvalidSyntaxException e) {
		throw new AssertionError(e);
	}
}

/**
 * Creates a filter that is the and-concatenation of this and the other
 * filter(s)
 * 
 * @param other the other filter to concatenate this filter with
 * @param more  possbily more filters to concatenate
 * @return
 */
default Filter and(Filter other, Filter... more) {
	try {
		StringBuilder builder = new StringBuilder();
		builder.append(this);
		builder.append(other);
		for (Filter filter : more) {
			builder.append(more);
		}
		return FrameworkUtil.createFilter("(&" + builder + ")");
	} catch (InvalidSyntaxException e) {
		throw new AssertionError(e);
	}
}

/**
 * Creates a filter that is the or-concatenation of this and the other filter(s)
 * 
 * @param other the other filter to concatenate this filter with
 * @param more  possbily more filters to concatenate
 * @return
 */
default Filter or(Filter other, Filter... more) {
	try {
		StringBuilder builder = new StringBuilder();
		builder.append(this);
		builder.append(other);
		for (Filter filter : more) {
			builder.append(more);
		}
		return FrameworkUtil.createFilter("(&" + builder + ")");
	} catch (InvalidSyntaxException e) {
		throw new AssertionError(e);
	}
}

/**
 * @param key   the key
 * @param value the value
 * @return a Filter matching the given key and value
 */
static Filter eq(String key, String value) {
	Objects.requireNonNull(key, "key can't be null");
	Objects.requireNonNull(value, "value can't be null");
	if (key.isBlank()) {
		throw new IllegalArgumentException("key can't be blank");
	}
	try {
		return FrameworkUtil.createFilter("(" + key + "=" + value + ")");
	} catch (InvalidSyntaxException e) {
		throw new IllegalArgumentException(e);
	}
}
```
</details>

one can construct the examples from the Filter Javadoc like this:

```
// (cn=Babs Jensen)
Filter.eq("cn", "Babs Jensen"); 

// (!(cn=Tim Howes))
Filter.eq("cn", "Tim Howes").not(); 

// (&(" + Constants.OBJECTCLASS + "=Person)(|(sn=Jensen)(cn=Babs J*)))
Filter.eq(Constants.OBJECTCLASS, "Person") //
	.and(
		Filter.eq("sn", "Jensen").or(Filter.eq("cn", "Babs J*")
	)
); 
```

**Reply by HannesWell** [reply]
In general I find this a good suggestion. It could be considered to name the method `equal()` instead of `eq()` and to make not() a static method so that it can be read more human readable ( `not(equal(..))`).
Also both would be possible. For example `java.util.function.Predicate` has both, a default method `negate()` and a static method `not()`.

**Reply by HannesWell** [reply]
It could also be considered to have specialized equals for common cases like one for `objectclass` filter, that accepts a Class:
`hasClass(Person.class)` or `isOfClass(Person.class)`. At the moment that is often programmed like:
`"(objectclass=" + Persion.class.getSimpleName() + ")"`.
The method name should clarify that the filtered object has to be of that type and must not be of a sub-type (or is that possible as well? I have to admit I don't know that).

**Reply by laeubi** [reply]
The problem with static methods is that they can't be overridden by the implementation by a more efficient ways, so the static ones should really only be an entry point to get an initial filter and then operate on the filter object itself.

**Reply by HannesWell** [reply]
That's right. But if there are both, then the static `not(filter)` could just call `not/negate()` on the passed filter object and return the result. This way the default not/negate() method can be overriden, but could still have the syntactic sugar.

**Comment by laeubi** [comment]
I just came along an example that I think is high-level enough (so no "we want to discourage use of...) and probably involves programmatic use cases as well (so no "filter only created once"), and of wider audience (so no "is of rarely use") and even part of OSGi Spec already (so no "they better should use") and **uses a builder pattern to construct object** that is the [Repository Service Specification](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.repository.html).

Just some examples where non trivial filters are constructed, and even if they are static for demonstration purpose would often be constructed at runtime because of some programmatic input:

- [Using a Repository](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.repository.html#d0e78202) uses `String.Format()` to build a filter, but of course this is limited to exact parameter count and attribute names
- [Combining Requirements](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.repository.html#d0e78320) has complex search strings
- [RequirementBuilder](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.repository.html#org.osgi.service.repository.RequirementBuilder)

**Comment by stbischof** [comment]
@laeubi fyi: bnd has a FilterBuilder
https://github.com/bndtools/bnd/blob/bb753403d95ca250eeb31528e916e61135c5b6a2/biz.aQute.bndlib/src/aQute/bnd/osgi/resource/FilterBuilder.java

**Reply by laeubi** [reply]
Thanks, probably this can be contributed to OSGi then so have it in core?
